Add share buttons at the top of each blog post

grmbyrn opened the pull request:
## Summary

This pull request introduces a new SvelteKit component named `ShareButton.svelte`, designed to enable users to easily share blog articles on various social media platforms. Additionally, it integrates the new component into the existing `+layout.svelte` to provide users with sharing options for individual blog posts.

## Changes Made

### New Component: `ShareButton.svelte`
- The new component includes a list of social media links, each represented by an icon, allowing users to share the blog post on platforms such as LinkedIn, Facebook, Twitter, and via email.
- The component receives the `slug` of the blog post as a prop to generate the correct shareable URLs.

### Integration into `+layout.svelte`
- The `ShareButton` component has been integrated into the `+layout.svelte` file. This ensures that every individual blog post now includes social media sharing options.
- The `slug` is passed to the `ShareButton` component, making it specific to each blog post.

Commit:
Add share buttons at the top of each blog post

grmbyrn commented:
Hi @josecelano, have you had a chance to look at this PR?

josecelano reviewed (changes requested):
Hi @grmbyrn,

- The code looks good.
- The design is OK. They are the way I would expect them to be and to look like.
- Some of them are not working, at least locally.

### LinkedIn 🆗 

![image](https://github.com/torrust/torrust-website/assets/58816/18885278-032f-4ee9-9ac8-1ede78a7c110)


### Facebook 🆗 

![image](https://github.com/torrust/torrust-website/assets/58816/282b78c7-28c7-490c-a943-22ec1b61482e)

### X

It seems it only adds the link but it could be the way it works. I was expecting the same behavior as LinkedIn and Facebook

![image](https://github.com/torrust/torrust-website/assets/58816/5139b128-51f7-4923-8ef0-d4120bfc9ef6)

### Email

![image](https://github.com/torrust/torrust-website/assets/58816/e1c994f9-8574-43f3-aef5-197cb82c69d1)

In my case, it's not adding the link or title. But It could be my email app.

I see that the parameters in the URL are not encoded.

```
mailto:?subject=I wanted you to see this blog post&body=Check out this blog post https://torrust.com/${slug}
```

grmbyrn committed:
Update ShareButton.svelte

* Encoded parameters in URL.
* Added title to tweet message.

grmbyrn commented:
Hi @josecelano, thanks for looking at that.

With the image on Twitter, I tried a number of things such as editing the URL and looking into meta tags, but I noticed that some other sites had the same reaction when I shared them:
<img width="594" alt="Screenshot 2024-01-14 at 23 56 01" src="https://github.com/torrust/torrust-website/assets/95353365/e9954917-6513-473d-8ae5-47c9421763be">

I added the title of the article at least so there would be more info outside of just the link:
<img width="608" alt="Screenshot 2024-01-14 at 23 54 27" src="https://github.com/torrust/torrust-website/assets/95353365/d889f7a8-093e-46fe-9cc4-a5a2bfa84269">

Also, I edited a bit how the email looked when sent. I sent it to another email address I have and it arrived like this:
<img width="657" alt="Screenshot 2024-01-14 at 23 53 30" src="https://github.com/torrust/torrust-website/assets/95353365/c254b5cc-789c-4159-8a2d-5437d3f20152">

I also encoded the URLs to look like this example:
```
mailto:?subject=${encodedSubject}&body='${title}' is a really interesting blog post from Torrust. Check it out here: ${siteBaseUrl}/${encodedSlug}
```

josecelano commented:
Hi @grmbyrn, 

- http://www.linkedin.com/shareArticle?mini=true&url=https://torrust.com/deploying-torrust-to-production

- https://www.facebook.com/sharer.php?u=https://torrust.com/deploying-torrust-to-production

- https://twitter.com/share?url=https://torrust.com/deploying-torrust-to-production&text=undefined

- mailto:?subject=I%20wanted%20you%20to%20see%20this%20blog%20post&amp;body='undefined' is a really interesting blog post from Torrust. Check it out here: https://torrust.com/deploying-torrust-to-production

For X and mailto, there is an "undefined". In "mailto" I think you have to use "&" not ";" (not sure). The body is not URL encoded.

https://datatracker.ietf.org/doc/html/rfc6068#section-6.1

grmbyrn committed:
Update ShareButton.svelte

Encoded body and title

grmbyrn commented:
> For X and mailto, there is an "undefined". In "mailto" I think you have to use "&" not ";" (not sure). The body is not URL encoded.
> 
> https://datatracker.ietf.org/doc/html/rfc6068#section-6.1

Do you mean "&" not "%"? I don't see any ";" in the mailto.

Also, I've tried encoding the title, as well as the body, to make it appear as it should rather than as undefined. It appears when I retweet or send the email as in the screenshots I sent before.

josecelano commented:
> > For X and mailto, there is an "undefined". In "mailto" I think you have to use "&" not ";" (not sure). The body is not URL encoded.
> > https://datatracker.ietf.org/doc/html/rfc6068#section-6.1
> 
> Do you mean "&" not "%"? I don't see any ";" in the mailto.
> 
> Also, I've tried encoding the title, as well as the body, to make it appear as it should rather than as undefined. It appears when I retweet or send the email as in the screenshots I sent before.

This is the mailto value when I checked it this morning:

```
mailto:?subject=I%20wanted%20you%20to%20see%20this%20blog%20post&amp;body='undefined' is a really interesting blog post from Torrust. Check it out here: https://torrust.com/deploying-torrust-to-production
```

There is a `;` before `body`.

grmbyrn commented:
> > > For X and mailto, there is an "undefined". In "mailto" I think you have to use "&" not ";" (not sure). The body is not URL encoded.
> > > https://datatracker.ietf.org/doc/html/rfc6068#section-6.1
> > 
> > 
> > Do you mean "&" not "%"? I don't see any ";" in the mailto.
> > Also, I've tried encoding the title, as well as the body, to make it appear as it should rather than as undefined. It appears when I retweet or send the email as in the screenshots I sent before.
> 
> This is the mailto value when I checked it this morning:
> 
> ```
> mailto:?subject=I%20wanted%20you%20to%20see%20this%20blog%20post&amp;body='undefined' is a really interesting blog post from Torrust. Check it out here: https://torrust.com/deploying-torrust-to-production
> ```
> 
> There is a `;` before `body`.

I have &body in the mailto, as well as each part encoded. Is there another reason why this is coming up?
```
`mailto:?subject=${encodedSubject}&body=${encodedTitle}${encodedBody}${siteBaseUrl}/${encodedSlug}`
```

josecelano commented:
> > > > For X and mailto, there is an "undefined". In "mailto" I think you have to use "&" not ";" (not sure). The body is not URL encoded.
> > > > https://datatracker.ietf.org/doc/html/rfc6068#section-6.1
> > > 
> > > 
> > > Do you mean "&" not "%"? I don't see any ";" in the mailto.
> > > Also, I've tried encoding the title, as well as the body, to make it appear as it should rather than as undefined. It appears when I retweet or send the email as in the screenshots I sent before.
> > 
> > 
> > This is the mailto value when I checked it this morning:
> > ```
> > mailto:?subject=I%20wanted%20you%20to%20see%20this%20blog%20post&amp;body='undefined' is a really interesting blog post from Torrust. Check it out here: https://torrust.com/deploying-torrust-to-production
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > There is a `;` before `body`.
> 
> I have &body in the mailto, as well as each part encoded. Is there another reason why this is coming up?
> 
> ```
> `mailto:?subject=${encodedSubject}&body=${encodedTitle}${encodedBody}${siteBaseUrl}/${encodedSlug}`
> ```

Hi @grmbyrn I've just realized the `;` belongs to the URL encoded `&`, which is `&amp;`

I think you are double-encoding. It should be something like:

```
mailto:majordomo@example.com?subject=aa&body=subscribe%20bamboo-l
```

But anyway, it seems it's not working with my email app. I've tested it with these examples:

https://www.scottseverance.us/mailto.html

grmbyrn committed:
Update ShareButton.svelte

Encoded all of body into one encodeURIComponent.

grmbyrn commented:
> > > > > For X and mailto, there is an "undefined". In "mailto" I think you have to use "&" not ";" (not sure). The body is not URL encoded.
> > > > > https://datatracker.ietf.org/doc/html/rfc6068#section-6.1
> > > > 
> > > > 
> > > > Do you mean "&" not "%"? I don't see any ";" in the mailto.
> > > > Also, I've tried encoding the title, as well as the body, to make it appear as it should rather than as undefined. It appears when I retweet or send the email as in the screenshots I sent before.
> > > 
> > > 
> > > This is the mailto value when I checked it this morning:
> > > ```
> > > mailto:?subject=I%20wanted%20you%20to%20see%20this%20blog%20post&amp;body='undefined' is a really interesting blog post from Torrust. Check it out here: https://torrust.com/deploying-torrust-to-production
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > There is a `;` before `body`.
> > 
> > 
> > I have &body in the mailto, as well as each part encoded. Is there another reason why this is coming up?
> > ```
> > `mailto:?subject=${encodedSubject}&body=${encodedTitle}${encodedBody}${siteBaseUrl}/${encodedSlug}`
> > ```
> 
> Hi @grmbyrn I've just realized the `;` belongs to the URL encoded `&`, which is `&amp;`
> 
> I think you are double-encoding. It should be something like:
> 
> ```
> mailto:majordomo@example.com?subject=aa&body=subscribe%20bamboo-l
> ```
> 
> But anyway, it seems it's not working with my email app. I've tested it with these examples:
> 
> https://www.scottseverance.us/mailto.html

Hi @josecelano I encoded the entire body into one `encodeURIComponent`. Maybe this fixes the double coding.

I've checked through a few different emails and it's sending me the email as expected, searched for reasons behind this, and asked ChatGPT but I don't know why it's not working with your email app. Is there anything else we can do to move this along?

josecelano commented:
> > > > > > For X and mailto, there is an "undefined". In "mailto" I think you have to use "&" not ";" (not sure). The body is not URL encoded.
> > > > > > https://datatracker.ietf.org/doc/html/rfc6068#section-6.1
> > > > > 
> > > > > 
> > > > > Do you mean "&" not "%"? I don't see any ";" in the mailto.
> > > > > Also, I've tried encoding the title, as well as the body, to make it appear as it should rather than as undefined. It appears when I retweet or send the email as in the screenshots I sent before.
> > > > 
> > > > 
> > > > This is the mailto value when I checked it this morning:
> > > > ```
> > > > mailto:?subject=I%20wanted%20you%20to%20see%20this%20blog%20post&amp;body='undefined' is a really interesting blog post from Torrust. Check it out here: https://torrust.com/deploying-torrust-to-production
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > There is a `;` before `body`.
> > > 
> > > 
> > > I have &body in the mailto, as well as each part encoded. Is there another reason why this is coming up?
> > > ```
> > > `mailto:?subject=${encodedSubject}&body=${encodedTitle}${encodedBody}${siteBaseUrl}/${encodedSlug}`
> > > ```
> > 
> > 
> > Hi @grmbyrn I've just realized the `;` belongs to the URL encoded `&`, which is `&amp;`
> > I think you are double-encoding. It should be something like:
> > ```
> > mailto:majordomo@example.com?subject=aa&body=subscribe%20bamboo-l
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But anyway, it seems it's not working with my email app. I've tested it with these examples:
> > https://www.scottseverance.us/mailto.html
> 
> Hi @josecelano I encoded the entire body into one `encodeURIComponent`. Maybe this fixes the double coding.
> 
> I've checked through a few different emails and it's sending me the email as expected, searched for reasons behind this, and asked ChatGPT but I don't know why it's not working with your email app. Is there anything else we can do to move this along?

hi, @grmbyrn don't worry about not working with my email app. It could be my email app does not allow adding a body. However, it was still not working for some reasons:

1. Missing injected title from parent component
2. Undefined in the URL  (the missing title)
3. The `&`in the URL was still encoded to `&amp;`. If you use directly a string Svelte makes sure it's encoded but in this case is already encoded. I have fixed the problem in this PR:
 
https://github.com/torrust/torrust-website/pull/130

grmbyrn committed:
Update ShareButton.svelte

Committed recommended change: Function to prevent Svelte from escaping any HTML entities. Parameters are already encoded. And we don't want to encode valid URL chars like the "&".

grmbyrn commented:
Recommended change from [#130](https://github.com/torrust/torrust-website/pull/130) committed.

josecelano commented:
ACK 9f7c4125942dee7836069488252bdb1dcc4f344e
